networking/configuring-a-custom-pki.adoc

```diff
@@ -0,0 +1,56 @@
+[id="configuring-a-custom-pki"]
+= Configuring a custom PKI
+include::modules/common-attributes.adoc[]
+:context: configuring-a-custom-pki
+
+toc::[]
+
+Some platform components (i.e., web console) use Routes for communication and,
+therefore, must trust other components' certificates to interact with them. If
+you're using a custom public key infrastructure (PKI), you must configure it so
+its privately signed CA certificates are recognized across the cluster.
+
+You can leverage the Proxy API to configure your custom PKI. You must do this
+either during installation *or* at runtime.
+
+* During *installation*, configure the cluster-wide proxy. You must define your
+privately signed CA certificates in the `install-config.yaml` file's
+`additionalTrustBundle` setting.
++
+The installation program generates a ConfigMap that is named `user-ca-bundle`
+that contains the additional CA certificates you defined. The Cluster Network
+Operator then creates a `trusted-ca-bundle` ConfigMap that merges these contents
+with the {op-system-first} trust bundle; this ConfigMap is referenced in the
+Proxy object's `trustedCA` field.
+
+* At *runtime*, enable the cluster-wide proxy. This involves creating a
+ConfigMap that contains the privately signed CA certificates that should be
+trusted by the cluster, and then modifying the proxy resource with the
+`trustedCA` referencing the privately signed certificates' ConfigMap.
+
+[NOTE]
+====
+The installer configuration's `additionalTrustBundle` field and the proxy
+resource's `trustedCA` field are used to manage the cluster-wide trust bundle;
+`additionalTrustBundle` is used at install time and `trustedCA` is used at run
+time.
+
+The `trustedCA` field is a reference to a `ConfigMap` containing the custom
+certificate and key pair used by the cluster component.
+====
+
+include::modules/installation-configure-proxy.adoc[leveloffset=+1]
+
+include::modules/nw-proxy-configure-object.adoc[leveloffset=+1]
+
+== Checkpoint
+
+You have defined your privately signed CA certificates in one of two ways:
+
+* In the `install-config.yaml` file's `additionalTrustBundle` field
+  (installation).
+
+* In the `user-ca-bundle.yaml` file's `ca-bundle.crt` field (runtime). Then you
```

Comment:
It might be helpful to show examples of the install and runtime configs:

__Runtime Config__

1. configmap config:
```
apiVersion: v1
data:
  ca-bundle.crt: |
    -----BEGIN CERTIFICATE-----
  <SNIP>
    -----END CERTIFICATE-----
kind: ConfigMap
metadata:
  name: user-ca-bundle
  namespace: openshift-config
```

2. proxy config:
```
apiVersion: config.openshift.io/v1
kind: Proxy
metadata:
  name: cluster
spec:
  trustedCA:
    name: user-ca-bundle
status:
```

__ Install time__
```
$ cat install-config.yaml 
<SNIP>
additionalTrustBundle: |
    -----BEGIN CERTIFICATE-----
<SNIP>
    -----END CERTIFICATE-----
```

Reply:
The config examples are actually spelled out in the modules I included in the assembly:
- [Runtime](http://file.rdu.redhat.com/~choag/112019/ingress-cert/networking/configuring-a-custom-pki.html#nw-proxy-configure-object_configuring-a-custom-pki)
- [Install](http://file.rdu.redhat.com/~choag/112019/ingress-cert/networking/configuring-a-custom-pki.html#installation-configure-proxy_configuring-a-custom-pki)

I added the *Checkpoint* to specifically call out the fields directly related to the additional CA certs. I included the additional proxy information because that seems important to the context of the proxy API we're leveraging here. Does that make sense?

Reply:
@danehans Or are you suggesting I call those examples out *again* in the checkpoint?

Reply:
> I added the Checkpoint to specifically call out the fields directly related to the additional CA certs. I included the additional proxy information because that seems important to the context of the proxy API we're leveraging here. Does that make sense?

OK, I missed that.

Reply:
no worries; thanks for reviewing!